relayer/src/link.rs

```diff
@@ -411,45 +442,95 @@ impl RelayPath {
         Err(LinkError::OldPacketClearingFailed)
     }
 
+    /// Should not run more than once per execution.
     pub fn clear_packets(&mut self, height: Height) -> Result<(), LinkError> {
         if self.clear_packets {
-            self.src_height = height
-                .decrement()
-                .map_err(|e| LinkError::Failed(e.to_string()))?;
-
-            self.relay_pending_packets()?;
+            self.relay_pending_packets(height)?;
             self.clear_packets = false;
         }
 
         Ok(())
     }
 
     /// Iterate through the IBC Events, build the message for each and collect all at same height.
-    /// Send a multi message transaction with these, prepending the client update
     pub fn relay_from_events(&mut self, batch: EventBatch) -> Result<(), LinkError> {
         self.clear_packets(batch.height)?;
 
-        // collect relevant events in self.all_events
+        // Collect relevant events in `self.all_events`.
         self.collect_events(&batch.events);
         self.adjust_events_height()?;
 
+        if self.channel.connection_delay.as_nanos() > 0 {
+            self.relay_with_delay()
+        } else {
+            self.relay_multi_msg()
+        }
+    }
+
+    /// Sends the client update in a separate transaction from packet messages, allowing for the
+    /// delay period to pass between the client update and packet messages.
+    fn relay_with_delay(&mut self) -> Result<(), LinkError> {
+        // If there are some new events, update src. and dest. clients, then schedule the relaying
+        // of packets corresponding to these events.
+        if !self.all_events.is_empty() {
+            let dst_update_height = self.dst_latest_height()?;
+            let src_update_height = self.all_events[0].height().increment();
+            self.update_clients(src_update_height, dst_update_height)?;
+
+            // Schedule the event batch, and let it sit for a while until the delay period passes
+            self.schedule_batch(self.all_events.clone(), dst_update_height);
+        }
+
+        // Iterate through all the outstanding event batches that have fulfilled their delay period,
+        // and send the packet messages corresponding to these events
+        while let Some(batch) = self.try_fetch_scheduled_batch() {
+            self.all_events = batch.events;
+
```

Comment:
The problem is that we look at scheduled batches on some IBC event or new block event. For the `NewBlock` we depend on the block time which varies a lot. For example on `cosmoshub-4` you can see even >60sec times. This means that the delay will in fact be higher than the particular next block time after the event height.

Reply:
fixed